src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/TranscodingReadStream.cs

```diff
@@ -0,0 +1,244 @@
+                    return 0;
+                }
+            }
+
+            bool completed = false;
+            int charsRead = default;
+            int bytesWritten = default;
+            // If the destination buffer is smaller than GetMaxByteCount(1), we avoid encoding to the destination and we use the overflow buffer instead.
+            if (readBuffer.Count > OverflowBufferSize || _charBuffer.Count == 0)
+            {
+                _encoder.Convert(_charBuffer.Array!, _charBuffer.Offset, _charBuffer.Count, readBuffer.Array!, readBuffer.Offset, readBuffer.Count,
+                    flush: false, out charsRead, out bytesWritten, out completed);
+            }
+
+            _charBuffer = _charBuffer.Slice(charsRead);
+
+            if (completed)
+            {
+                return bytesWritten;
+            }
+            else
+            {
+                if (bytesWritten > 0)
+                {
+                    return bytesWritten;
+                }
+
+                _encoder.Convert(_charBuffer.Array!, _charBuffer.Offset, _charBuffer.Count, _overflowBuffer.Array!, 0, _overflowBuffer.Array!.Length,
+                    flush: false, out int overFlowChars, out int overflowBytes, out completed);
+
+                Debug.Assert(overflowBytes > 0 && overFlowChars > 0, "We expect writes to the overflow buffer to always succeed since it is large enough to accommodate at least one char.");
+
+                _charBuffer = _charBuffer.Slice(overFlowChars);
+
+                // readBuffer: [ 0, 0, ], overflowBuffer: [ 7, 13, 34, ]
+                // Fill up the readBuffer to capacity, so the result looks like so:
+                // readBuffer: [ 7, 13 ], overflowBuffer: [ 34 ]
+                Debug.Assert(readBuffer.Count < overflowBytes);
+                _overflowBuffer.Array.AsSpan(0, readBuffer.Count).CopyTo(readBuffer);
+
+                Debug.Assert(_overflowBuffer.Array != null);
+
+                _overflowBuffer = new ArraySegment<byte>(_overflowBuffer.Array, readBuffer.Count, overflowBytes - readBuffer.Count);
+
+                Debug.Assert(_overflowBuffer.Count > 0);
+
+                return readBuffer.Count;
+            }
+        }
+
+        private async Task<int> ReadInputChars(CancellationToken cancellationToken)
+        {
+            // If we had left-over bytes from a previous read, move it to the start of the buffer and read content into
+            // the segment that follows.
+            Debug.Assert(_byteBuffer.Array != null);
+            Buffer.BlockCopy(_byteBuffer.Array, _byteBuffer.Offset, _byteBuffer.Array, 0, _byteBuffer.Count);
+
+            int offset = _byteBuffer.Count;
+            int count = _byteBuffer.Array.Length - _byteBuffer.Count;
+
+            int bytesRead = await _stream.ReadAsync(_byteBuffer.Array, offset, count, cancellationToken).ConfigureAwait(false);
```

Comment:
Using the memory-based overload here would be useful, too, and actually more so than with the write stream.